test/unittests/execute_floating_point_test.cpp

```diff
@@ -66,4 +69,326 @@ TEST(execute_floating_point, f64_add)
+        EXPECT_THAT(execute(*instance, 0, {signaling_nan}), Traps());
+        EXPECT_THAT(execute(*instance, 0, {-signaling_nan}), Traps());
+
+        const auto inf = FloatLimits::infinity();
+        EXPECT_THAT(execute(*instance, 0, {inf}), Traps());
+        EXPECT_THAT(execute(*instance, 0, {-inf}), Traps());
+
+        EXPECT_THAT(execute(*instance, 0, {FloatLimits::max()}), Traps());
+        EXPECT_THAT(execute(*instance, 0, {FloatLimits::lowest()}), Traps());
+    }
+
+    if constexpr (IntLimits::is_signed)
+    {
+        // Something bigger than -2.0.
+        const auto arg = std::nextafter(FloatT{-2}, FloatT{0});
+        const auto result = execute(*instance, 0, {arg});
+        EXPECT_EQ(result.value.template as<IntT>(), FloatT{-1});
+    }
+}
+
+
+template <typename T>
+class execute_floating_point_convert : public testing::Test
+{
+};
+
+using ConvertPairs = testing::Types<ConversionPair<int32_t, float>, ConversionPair<uint32_t, float>,
+    ConversionPair<int64_t, float>, ConversionPair<uint64_t, float>,
+    ConversionPair<int32_t, double>, ConversionPair<uint32_t, double>,
+    ConversionPair<int64_t, double>, ConversionPair<uint64_t, double>>;
+TYPED_TEST_SUITE(execute_floating_point_convert, ConvertPairs, ConversionName);
+
+TYPED_TEST(execute_floating_point_convert, convert)
+{
+    using IntT = typename TypeParam::src_type;
+    using FloatT = typename TypeParam::dst_type;
+    using IntLimits = std::numeric_limits<IntT>;
+    using FloatLimits = std::numeric_limits<FloatT>;
+
+    /* wat2wasm
+    (func (param i32) (result f32)
+      local.get 0
+      f32.convert_i32_s
+    )
+    */
+    auto wasm = from_hex("0061736d0100000001060160017f017d030201000a070105002000b20b");
+
+    // Find and replace changeable values: types and the conversion instruction.
+    auto& param_type = wasm[wasm.find(static_cast<uint8_t>(ValType::i32))];
+    auto& result_type = wasm[wasm.find(static_cast<uint8_t>(ValType::f32))];
+    auto& opcode = wasm[wasm.find(static_cast<uint8_t>(Instr::f32_convert_i32_s))];
+    param_type = static_cast<uint8_t>(TypeParam::src_valtype);
+    result_type = static_cast<uint8_t>(TypeParam::dst_valtype);
+    opcode = static_cast<uint8_t>(TypeParam::opcode);
+
+    auto instance = instantiate(parse(wasm));
+
+    EXPECT_THAT(execute(*instance, 0, {IntT{0}}), Result(FloatT{0}));
+    EXPECT_THAT(execute(*instance, 0, {IntT{1}}), Result(FloatT{1}));
+
+    const auto max = IntLimits::max();
+    const auto exact = IntLimits::digits < FloatLimits::digits;
+    const auto max_expected = std::pow(FloatT{2}, FloatT{IntLimits::digits}) - FloatT{exact};
+    EXPECT_THAT(execute(*instance, 0, {max}), Result(max_expected));
+
+    if constexpr (IntLimits::is_signed)
+    {
+        static_assert(std::is_same_v<decltype(-max), IntT>);
+        EXPECT_THAT(execute(*instance, 0, {-max}), Result(-max_expected));
+
+        const auto min_expected = -std::pow(FloatT{2}, FloatT{IntLimits::digits});
```

Comment:
I wonder how `std::pow` is implemented, is it guaranteed to always have the exact result for power of 2?

Reply:
I don't think `pow()` matters here, but floating point types can represent all powers of 2 around integer ranges due to having the exponent part which produces `2^e` values.